### PR TITLE
Smart G4 magnetic field manager

### PR DESCRIPTION
**Pull request by civanch**
A new Geant4 field manager is added allowing to switch a set magnetic field parameters depending on track type and energy. This is needed because high precision tracking in field for low-energy particles is not needed in all CMS sub-detectors.

In the current variant of code this mechanism introduced but parameters are set to zero, so the simulation should produce unchanged results. New parameters of the method are configured by pythin and should be established after optimisation.

Additionally some clean-up is introduced:
  - improved Info printouts;
  - do not create field in master thread for ordinary runs - only in worker threads;
  - reduced number of objects registered in tls;
  - reduced number of mutable class members.

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
@davidlange6 , this PR would useful to integrate for this pre-release. Results for SIM hits production should be unchanged, because possible optimisation of tracking in field in this PR is disabled, so the field is handled as before.

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 1]
```diff
@@ -62,7 +62,7 @@
         NodeNames = cms.vstring('World')
     ),
     G4Commands = cms.vstring(),
-    SteppingVerbosity = cms.int32(0),
+    SteppingVerbosity = cms.int32(1),
```
Comment (davidlange6) [inline SimG4Core/Application/python/g4SimHits_cfi.py]: Hi @civanch - we don't want this change I think?

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by civanch** [comment]
@davidlange6 , sorry, verbosity level is fixed now.

**Comment by davidlange6** [comment]
+1

**Pull request merged**